tket/src/WeightSubgrMono/GraphTheoretic/NeighboursData.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/GraphTheoretic/NeighboursData.hpp"
+
+#include <algorithm>
+#include <set>
+#include <stdexcept>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+NeighboursData::NeighboursData() : m_number_of_edges(0) {}
+
+NeighboursData::NeighboursData(const GraphEdgeWeights& edges_and_weights) {
+  initialise(edges_and_weights);
+}
+
+std::size_t NeighboursData::get_number_of_edges() const {
+  return m_number_of_edges;
+}
+
+std::vector<VertexWSM> NeighboursData::get_neighbours_expensive(
+    VertexWSM v) const {
+  std::vector<VertexWSM> result;
+  const auto& neighbours_and_weights = get_neighbours_and_weights(v);
+  result.reserve(neighbours_and_weights.size());
+  for (const auto& entry : neighbours_and_weights) {
+    result.push_back(entry.first);
+  }
+  return result;
+}
+
+std::size_t NeighboursData::get_number_of_nonisolated_vertices() const {
+  return m_neighbours_and_weights.size();
+}
+
+void NeighboursData::initialise(const GraphEdgeWeights& edges_and_weights) {
+  // Only the edges (v1,v2) with v1<v2.
+  std::set<std::pair<VertexWSM, VertexWSM>> ordered_edges_seen;
+  std::set<VertexWSM> vertices_seen;
+  for (const auto& entry : edges_and_weights) {
+    const auto& v1 = entry.first.first;
+    const auto& v2 = entry.first.second;
+    vertices_seen.insert(v1);
+    vertices_seen.insert(v2);
+    if (v1 == v2) {
+      throw std::runtime_error("Loop found in graph; not allowed");
+    }
+    ordered_edges_seen.insert(get_edge(v1, v2));
+  }
+  if (vertices_seen.empty()) {
+    throw std::runtime_error("No edges passed to NeighboursData");
+  }
+  if (*vertices_seen.cbegin() != 0 ||
+      *vertices_seen.crbegin() != vertices_seen.size() - 1) {
```

Comment:
Same again; we're checking input data for validity. If later, we were to process output data, returned from this class, then an assert WOULD be justified, because it would be after the input data had been accepted.